core/network/src/protocol/sync.rs

```diff
@@ -1009,79 +1042,42 @@ impl<B: BlockT> ChainSync<B> {
 		match self.block_announce_validator.validate(&header, assoc_data) {
 			Ok(Validation::Success) => (),
 			Ok(Validation::Failure) => {
-				debug!(target: "sync", "block announcement validation of block {} from {} failed", hash, who);
+				debug!(target: "sync", "Block announcement validation of block {} from {} failed", hash, who);
 				return OnBlockAnnounce::Nothing
 			}
 			Err(e) => {
-				error!(target: "sync", "block announcement validation errored: {}", e);
+				error!(target: "sync", "Block announcement validation errored: {}", e);
 				return OnBlockAnnounce::Nothing
 			}
 		}
 
-		// stale block case
-		let requires_additional_data = !self.role.is_light();
-		if number <= self.best_queued_number {
-			if !(known_parent || self.is_already_downloading(header.parent_hash())) {
-				let block_status = self.client.block_status(&BlockId::Number(*header.number()))
-					.unwrap_or(BlockStatus::Unknown);
-				if block_status == BlockStatus::InChainPruned {
-					trace!(
-						target: "sync",
-						"Ignored unknown ancient block announced from {}: {} {:?}", who, hash, header
-					);
-					return OnBlockAnnounce::Nothing
-				}
-				trace!(
-					target: "sync",
-					"Considering new unknown stale block announced from {}: {} {:?}", who, hash, header
-				);
-				if let Some(request) = self.download_unknown_stale(&who, &hash) {
-					if requires_additional_data {
-						return OnBlockAnnounce::Request(who, request)
-					} else {
-						return OnBlockAnnounce::ImportHeader
-					}
-				} else {
-					return OnBlockAnnounce::Nothing
-				}
-			} else {
-				if ancient_parent {
-					trace!(target: "sync", "Ignored ancient stale block announced from {}: {} {:?}", who, hash, header);
-					return OnBlockAnnounce::Nothing
-				}
-				if let Some(request) = self.download_stale(&who, &hash) {
-					if requires_additional_data {
-						return OnBlockAnnounce::Request(who, request)
-					} else {
-						return OnBlockAnnounce::ImportHeader
-					}
-				} else {
-					return OnBlockAnnounce::Nothing
-				}
-			}
+		let requires_additional_data = !self.role.is_light() || !known_parent;
+		if !requires_additional_data {
+			trace!(target: "sync", "Importing new header announced from {}: {} {:?}", who, hash, header);
+			return OnBlockAnnounce::ImportHeader
 		}
 
-		if ancient_parent {
+		if ancient_parent && !requires_additional_data {
```

Comment:
There's a `if !requires_additional_data {}` right above - so either this `if` should be moved before the other one, or we can merge both in single one with `if ancient_parent { OnBlockAnnounce::Nothing } else { OnBlockAnnounce::ImportHeader }`. At the same time, I've noticed that currently on light client `fn block_status()` never returns `InChainPruned`. I'll prepare a follow-up PR that returns `InChainPruned` for headers that have been replaced by CHT.

Reply:
I suppose there should be a distinction between pruned state and pruned block here. For now the code treats any kind of pruning as a no go for import.